-   [Background](#background)
-   [The gwascat package for the EMBL-EBI (formerly NHGRI) GWAS
    catalog](#the-gwascat-package-for-the-embl-ebi-formerly-nhgri-gwas-catalog)
    -   [Basic operations, fields, and interactive
        tabulation](#basic-operations-fields-and-interactive-tabulation)
-   [GRASP](#grasp)
-   [Genomic contexts and interpretations of
    variants](#genomic-contexts-and-interpretations-of-variants)
    -   [Presence in exons](#presence-in-exons)
    -   [SIFT scores](#sift-scores)
    -   [ChromHmm segmentation](#chromhmm-segmentation)
    -   [Regions of chromatin
        modification](#regions-of-chromatin-modification)
-   [Conclusions](#conclusions)
-   [Appendix: Bioconductor infrastructure supporting genetic data
    analysis](#appendix-bioconductor-infrastructure-supporting-genetic-data-analysis)
    -   [Reference builds of the human genome
        sequence](#reference-builds-of-the-human-genome-sequence)
    -   [From dbSNP to GRanges](#from-dbsnp-to-granges)

In [1]:
## This code chunk was hidden in the original document, but was executed in the background
knitr::opts_chunk$set(results="hide", message=FALSE, warning=FALSE, fig.show="hide", echo=TRUE)

In [2]:
## This code chunk was hidden in the original document, but was executed in the background
suppressPackageStartupMessages({
library(BiocStyle)
library(AnnotationHub)
ah = AnnotationHub()
library(gwascat)
library(GenomicFiles)
library(rtracklayer)
library(DT)
library(SIFT.Hsapiens.dbSNP132)
library(grasp2db)
library(BSgenome)
library("SNPlocs.Hsapiens.dbSNP144.GRCh37")
#library(BSgenome.Hsapiens.NCBI.GRCh38)
library(BSgenome.Hsapiens.UCSC.hg19)
})

snapshotDate(): 2015-12-29
Now getting the GODb Object directly
Now getting the OrgDb Object directly
Now getting the TxDb Object directly


Background
==========

The table of contents of Vogel and Motulsky's [*Human Genetics: Problems and Approaches*](https://books.google.com/books?id=xuztCAAAQBAJ&lpg=PA6&dq=human%20genetics&pg=PR32#v=onepage&q=human%20genetics&f=false) is a worthy survey of concepts addressed in research on human genetics and genetic medicine. The frontiers of knowledge in the field are shifting, and expectations are high.

In this workflow, I aim to show how researchers can use R to interrogate important resources of use in human genetic epidemiology and medical genomics. I show how to program with two genome-wide association study (GWAS) catalogs, the [EMBL-EBI GWAS catalog](https://www.ebi.ac.uk/gwas/) and the [NHLBI GRASP v2.0](http://iapps.nhlbi.nih.gov/GRASP/Overview.aspx). Aspects of findings reported in these studies are then integrated with new functional and structural annotation resources to aid in variant interpretation. An appendix provides brief treatment of "reference genome builds" for *Homo sapiens*, packages for querying contents of the [NCBI dbSNP](http://www.ncbi.nlm.nih.gov/SNP/), and tools for obtaining and programming with gene models.

The gwascat package for the EMBL-EBI (formerly NHGRI) GWAS catalog
==================================================================

Basic operations, fields, and interactive tabulation
----------------------------------------------------

The NHGRI version of the GWAS catalog is presented using hg19( GRCh37) coordinates.

In [3]:
library(gwascat)
data(gwrngs19)
length(gwrngs19)
gwrngs19

[1] 17254

gwasloc instance with 17254 records and 35 attributes per record.
Extracted:  Mon Sep  8 13:08:13 2014 
Genome:  hg19 
Excerpt:
GRanges object with 5 ranges and 35 metadata columns:
    seqnames                 ranges strand | Date.Added.to.Catalog  PUBMEDID
       <Rle>              <IRanges>  <Rle> |           <character> <integer>
  1    chr19 [  7739177,   7739177]      * |            04/16/2014  24123702
  2     chr6 [ 32626601,  32626601]      * |            08/02/2014  24388013
  3     chr4 [ 38799710,  38799710]      * |            08/02/2014  24388013
  4     chr5 [110467499, 110467499]      * |            08/02/2014  24388013
  5     chr2 [102966549, 102966549]      * |            08/02/2014  24388013
    First.Author        Date                Journal
     <character> <character>            <character>
  1     Chung CM  03/03/2014 Diabetes Metab Res Rev
  2  Ferreira MA  12/30/2013 J Allergy Clin Immunol
  3  Ferreira MA  12/30/2013 J Allergy Clin Immunol
  4  Ferreira MA  1

While there are 17254 records, the number of unique loci is

In [4]:
length(unique(gwrngs19$SNPs))

[1] 13630

A full view of the metadata about each study result is available with the commands

``` r
library(DT)
datatable(as.data.frame(mcols(gwrngs19)), options=list(autoWidth=TRUE,
  style="height:30px"), pageLength=5)
```

The following command generates a table restricting attention to records related to asthma.

In [5]:
suppressWarnings({
aind = grep("sthma", gwrngs19$Disease.Trait)
easth = gwrngs19[aind]
datatable(as.data.frame(mcols(easth)), options=list(autoWidth=TRUE,
  style="height:30px", pageLength=5))
})

<!--

## Navigating traits using the EMBL-EBI Experimental Factor Ontology

Field `MAPPED_TRAIT_URI` includes a comma-delimited string with
URIs referring to an ontology for traits and other factors relevant
to biological experiments and observations.  The underlying
ontology is available in the form of an annotated algebraic graph.

``` r
data(efo.obo.g)
efo.obo.g
```

There are over 16000 terms in the ontology.  Terms and term-related
metadata are manipulated using methods of the *[graph](http://bioconductor.org/packages/graph)*
package.

``` r
nodes(efo.obo.g)[1:4] # imported directly from OBO
names(nodeData(efo.obo.g)[[1]])
sapply(nodeData(efo.obo.g)[1:4], "[[", "name")
```

Let's obtain the EFO annotation for SNP `rs347412`.

``` r
ind = which(ebicat38$SNPS == "rs347412")
urs = ebicat38$MAPPED_TRAIT_URI[ind]
urs
```

These entries must be converted to match the EFO OBO node
nomenclature.  We then find the EFO names of the factors annotated
to this SNP.

``` r
nn = uri2node(urs)
nd = nodeData(efo.obo.g, nn)
sapply(nd, "[[", "name")
```

The current representation of the ontology is a directed graph
with links pointing from a term to its semantic parent.  We
convert to an undirected graph to explore semantic neighborhoods of terms.
The `adj` method will return the nodes adjacent to a specified node.
Here we obtain the terms accessible from `respiratory system disease`
with a single step.

``` r
rsdn = adj(ugraph(efo.obo.g), "EFO:0000684")  # respiratory system disease
unlist(sapply(nodeData(efo.obo.g, rsdn[[1]]), "[[", "name"))
```

The *[RBGL](http://bioconductor.org/packages/RBGL)* package can be used to deploy diverse graph algorithms
against this ontology.

Once a node name of interest has been found, `node2uri` can be used
with code to find
GWAS hits deemed relevant by the curators.  We'll work with hg19
coordinates.

``` r
data(ebicat37)
library(GenomeInfoDb)
seqlevelsStyle(ebicat37) = "UCSC"
genome(ebicat37) = "hg19"
e270 = ebicat37[ grep(node2uri("EFO:0000270"), ebicat37$MAPPED_TRAIT_URI) ]
length(e270)
table(e270$DISEASE.TRAIT)[1:5]
```

-->
GRASP
=====

GRASP is a much denser catalog requiring a different approach to archiving and query resolution. Initial execution of `GRASP2()` will trigger a download of a 5GB SQLite database that can then be used with *[dplyr](http://cran.fhcrc.org/web/packages/dplyr/index.html)* programming. This download will not occur again unless the database has been centrally updated. This document does not evaluate the following chunk, but the output is precomputed and left static.

``` r
library(grasp2db)
v = tbl(GRASP2(), 'variant')
v %>% filter(Phenotype == "Asthma")
```

<pre><code>## Source: sqlite 3.8.6 [AnnotationHub()[[&quot;AH21414&quot;]]]
## From: variant [33,351 x 33]
## Filter: Phenotype == &quot;Asthma&quot; 
## 
##        NHLBIkey     PMID HUPfield SNPid_dbSNP134 chr_hg19  pos_hg19
## 1    2086050316 20860503 1/1/2014             18        7  11597475
## 2   20860503866 20860503 1/1/2014            535        9 138396251
## 3  208605031097 20860503 1/1/2014            686        5 174868700
## 4  208605031186 20860503 1/1/2014            699        1 230845794
## 5  208605031603 20860503 1/1/2014           1117        3  22085809
## 6  208605031980 20860503 1/1/2014           1320       22  22599537
## 7  208605032429 20860503 1/1/2014           1535       11  61597972
## 8  208605032734 20860503 1/1/2014           1695       11  67352689
## 9  208605032835 20860503 1/1/2014           1760        8    442079
## 10 208605033085 20860503 1/1/2014           1899       15  41689232
## ..          ...      ...      ...            ...      ...       ...
## Variables not shown: SNPidInPaper (chr), LocationWithinPaper (chr), Pvalue
##   (dbl), NegativeLog10PBin (int), Phenotype (chr), PlatformSNPsPassingQC
##   (chr), GWASancestryDescription (chr), InGene (chr), InLincRNA (chr),
##   InMiRNA (chr), InMiRNABS (chr), dbSNPfxn (chr), dbSNPMAF (chr),
##   dbSNPallelesHetSe (chr), dbSNPvalidation (int), dbSNPClinStatus (chr),
##   ORegAnno (chr), ConservPredTFBS (chr), HumanEnhancer (chr), RNAedit
##   (chr), PolyPhen2 (chr), SIFT (chr), LS_SNP (chr), UniProt (chr),
##   EqtlMethMetabStudy (int), DiscoverySampleDescription (chr),
##   ReplicationSampleDescription (chr)</code></pre>
Genomic contexts and interpretations of variants
================================================

Presence in exons
-----------------

We can map our GWAS hits to exons using the TxDb infrastructure.

In [6]:
library(TxDb.Hsapiens.UCSC.hg19.knownGene)
allex = exons(TxDb.Hsapiens.UCSC.hg19.knownGene)
subsetByOverlaps( easth, allex )

gwasloc instance with 12 records and 35 attributes per record.
Extracted:  Mon Sep  8 13:08:13 2014 
Genome:  hg19 
Excerpt:
GRanges object with 5 ranges and 35 metadata columns:
      seqnames                 ranges strand | Date.Added.to.Catalog  PUBMEDID
         <Rle>              <IRanges>  <Rle> |           <character> <integer>
    3     chr4 [ 38799710,  38799710]      * |            08/02/2014  24388013
    6    chr17 [ 38122680,  38122680]      * |            08/02/2014  24388013
  734     chr7 [105658451, 105658451]      * |            07/25/2014  24241537
  735    chr17 [ 38062196,  38062196]      * |            07/25/2014  24241537
  738    chr17 [ 38121993,  38121993]      * |            07/25/2014  24241537
      First.Author        Date                Journal
       <character> <character>            <character>
    3  Ferreira MA  12/30/2013 J Allergy Clin Immunol
    6  Ferreira MA  12/30/2013 J Allergy Clin Immunol
  734 Bonnelykke K  11/17/2013              Nat Gene

SIFT scores
-----------

We query the SIFT resource using dbSNP identifiers.

In [7]:
rsids = easth$SNPs
library(SIFT.Hsapiens.dbSNP132)
subst = c("RSID", "METHOD", "PREDICTION", "SCORE")
sif = AnnotationDbi::select(SIFT.Hsapiens.dbSNP132, keys=rsids, cols=subst)
datatable(na.omit(sif))

Warning message:
: 138 keys not found in SIFT database: rs9273373 rs1438673 ... rs16937883 rs7216389

ChromHmm segmentation
---------------------

We'll use the fetal lung sample from the epigenomics road map as provided by *[AnnotationHub](http://bioconductor.org/packages/AnnotationHub)*. We use prior knowledge that tag "E088" refers to the fetal lung tissue study.

In [8]:
library(AnnotationHub)
ah = AnnotationHub()
lq = AnnotationHub::query(ah, c("E088", "state"))
lq
cstates = subsetByOverlaps( ah[["AH46941"]], easth )
sort(table(cstates$name), decreasing=TRUE)

snapshotDate(): 2015-12-29


AnnotationHub with 1 record
# snapshotDate(): 2015-12-29 
# names(): AH46941
# $dataprovider: BroadInstitute
# $species: Homo sapiens
# $rdataclass: GRanges
# $title: E088_15_coreMarks_mnemonics.bed.gz
# $description: 15 state chromatin segmentations from EpigenomeRoadMap Project
# $taxonomyid: 9606
# $genome: hg19
# $sourcetype: BED
# $sourceurl: http://egg2.wustl.edu/roadmap/data/byFileType/chromhmmSegmenta...
# $sourcelastmodifieddate: 2013-10-11
# $sourcesize: 2825556
# $tags: EpigenomeRoadMap, chromhmmSegmentations, ChmmModels,
#   coreMarks, E088, Other, LNG.FET, Fetal Lung 
# retrieve record with 'object[["AH46941"]]' 

Quiescent/Low      Weak transcription Weak Repressed PolyComb 
                     65                      18                      17 
   Strong transcription               Enhancers      Repressed PolyComb 
                     11                       6                       6 
        Heterochromatin              Active TSS     Bivalent/Poised TSS 
                      3                       2                       2 
        Genic enhancers     ZNF genes & repeats 
                      2                       1

In this way we can label variants according to their tissue-specific epigenetic contexts.

Regions of chromatin modification
---------------------------------

We'll check for coincidence of our GWAS hits with peaks identified with H3K4me1 marks in fetal lung fibroblasts, using component AH43875 of the *[AnnotationHub](http://bioconductor.org/packages/AnnotationHub)*.

In [9]:
library(AnnotationHub)
ah = AnnotationHub()
h3kf = ah[["AH43875"]]
subsetByOverlaps(easth, h3kf)

snapshotDate(): 2015-12-29


gwasloc instance with 12 records and 35 attributes per record.
Extracted:  Mon Sep  8 13:08:13 2014 
Genome:  hg19 
Excerpt:
GRanges object with 5 ranges and 35 metadata columns:
       seqnames                 ranges strand | Date.Added.to.Catalog  PUBMEDID
          <Rle>              <IRanges>  <Rle> |           <character> <integer>
     9    chr15 [ 67468285,  67468285]      * |            08/02/2014  24388013
    11    chr16 [ 11228712,  11228712]      * |            08/02/2014  24388013
   416    chr19 [ 52532271,  52532271]      * |            07/25/2014  24280104
  2186     chr3 [ 35639826,  35639826]      * |            02/06/2014  23829686
  2193     chr3 [177855632, 177855632]      * |            02/06/2014  23829686
       First.Author        Date                Journal
        <character> <character>            <character>
     9  Ferreira MA  12/30/2013 J Allergy Clin Immunol
    11  Ferreira MA  12/30/2013 J Allergy Clin Immunol
   416        Wu AC  11/23/2013 J Allergy

Conclusions
===========

The use of *[GenomicRanges](http://bioconductor.org/packages/GenomicRanges)* infrastructure for representing sets of DNA variants leads to fairly simple merge and intersection operations based on genomic coordinates. These operations are useful for sorting variants into categories based on structural or functional modeling. Richly annotated ranges can be used to manage and program with GWAS catalogs, leading to efficient coupling of genomic assay results with findings of genetic epidemiology.

Appendix: Bioconductor infrastructure supporting genetic data analysis
======================================================================

Reference builds of the human genome sequence
---------------------------------------------

<!--
The most recent build of the human genomic sequence
is labeled GRCh38.  Using Bioconductor, we can be very concrete about what this
is.
-->
The second-to-last build of the human genomic sequence is labeled hg19. Using Bioconductor, we can be very concrete about what this is.

In [10]:
library(BSgenome.Hsapiens.UCSC.hg19)
class(Hsapiens)
Hsapiens
class(Hsapiens$"chr17")
Hsapiens$"chr17"

[1] "BSgenome"
attr(,"package")
[1] "BSgenome"

Human genome:
# organism: Homo sapiens (Human)
# provider: UCSC
# provider version: hg19
# release date: Feb. 2009
# release name: Genome Reference Consortium GRCh37
# 93 sequences:
#   chr1                  chr2                  chr3                 
#   chr4                  chr5                  chr6                 
#   chr7                  chr8                  chr9                 
#   chr10                 chr11                 chr12                
#   chr13                 chr14                 chr15                
#   ...                   ...                   ...                  
#   chrUn_gl000235        chrUn_gl000236        chrUn_gl000237       
#   chrUn_gl000238        chrUn_gl000239        chrUn_gl000240       
#   chrUn_gl000241        chrUn_gl000242        chrUn_gl000243       
#   chrUn_gl000244        chrUn_gl000245        chrUn_gl000246       
#   chrUn_gl000247        chrUn_gl000248        chrUn_gl000249       
# (use 'seqnames()' to see all the sequence name

[1] "DNAString"
attr(,"package")
[1] "Biostrings"

  81195210-letter "DNAString" instance
seq: AAGCTTCTCACCCTGTTCCTGCATAGATAATTGCAT...GTGGGTGTGGGTGTGGTGTGTGGGTGTGGGTGTGGT

From dbSNP to GRanges
---------------------

A number of packages represent snapshots of NCBI dbSNP.

In [11]:
library(BSgenome)
available.SNPs()

[1] "SNPlocs.Hsapiens.dbSNP.20090506"     
 [2] "SNPlocs.Hsapiens.dbSNP.20100427"     
 [3] "SNPlocs.Hsapiens.dbSNP.20101109"     
 [4] "SNPlocs.Hsapiens.dbSNP.20110815"     
 [5] "SNPlocs.Hsapiens.dbSNP.20111119"     
 [6] "SNPlocs.Hsapiens.dbSNP.20120608"     
 [7] "SNPlocs.Hsapiens.dbSNP141.GRCh38"    
 [8] "SNPlocs.Hsapiens.dbSNP142.GRCh37"    
 [9] "SNPlocs.Hsapiens.dbSNP144.GRCh37"    
[10] "SNPlocs.Hsapiens.dbSNP144.GRCh38"    
[11] "XtraSNPlocs.Hsapiens.dbSNP141.GRCh38"
[12] "XtraSNPlocs.Hsapiens.dbSNP144.GRCh37"
[13] "XtraSNPlocs.Hsapiens.dbSNP144.GRCh38"

Functions available for a recent build are:

In [12]:
library("SNPlocs.Hsapiens.dbSNP144.GRCh37")
ls(pos="package:SNPlocs.Hsapiens.dbSNP144.GRCh37")

[1] "COMPATIBLE_BSGENOMES"             "getSNPcount"                     
[3] "getSNPlocs"                       "rsid2alleles"                    
[5] "rsid2loc"                         "rsidsToGRanges"                  
[7] "SNPlocs.Hsapiens.dbSNP144.GRCh37"

We can retrieve data on a chromosome. Note the peculiar nomenclature for chromosomes used with dbSNP. The `seqlevelsStyle` methods of *[GenomeInfoDb](http://bioconductor.org/packages/GenomeInfoDb)* can be used to manage these nomenclatures systematically.

In [13]:
snpsBySeqname(SNPlocs.Hsapiens.dbSNP144.GRCh37, "ch20")

GRanges object with 2990255 ranges and 2 metadata columns:
            seqnames               ranges strand   |   RefSNP_id
               <Rle>            <IRanges>  <Rle>   | <character>
        [1]     ch20       [60039, 60039]      +   | rs772768480
        [2]     ch20       [60309, 60309]      +   | rs574852966
        [3]     ch20       [60343, 60343]      +   | rs527639301
        [4]     ch20       [60366, 60366]      +   | rs762720439
        [5]     ch20       [60419, 60419]      +   | rs538242240
        ...      ...                  ...    ... ...         ...
  [2990251]     ch20 [62965305, 62965305]      +   | rs577023641
  [2990252]     ch20 [62965324, 62965324]      +   | rs777323048
  [2990253]     ch20 [62965337, 62965337]      +   |   rs2760732
  [2990254]     ch20 [62965354, 62965354]      +   | rs546194182
  [2990255]     ch20 [62965384, 62965384]      +   |   rs2760733
            alleles_as_ambig
                 <character>
        [1]                Y
        [